src/charm.py

```diff
@@ -24,36 +29,73 @@ def __init__(self, *args: typing.Any):
             args: Arguments passed to the CharmBase parent constructor.
         """
         super().__init__(*args)
-        self.irc = IRCService()
+        self._irc = IRCBridgeService()
+        self._database = DatabaseObserver(self, DATABASE_RELATION_NAME)
+        self._matrix = MatrixObserver(self, MATRIX_RELATION_NAME)
         self.framework.observe(self.on.config_changed, self._on_config_changed)
         self.framework.observe(self.on.install, self._on_install)
         self.framework.observe(self.on.start, self._on_start)
         self.framework.observe(self.on.stop, self._on_stop)
-        self.framework.observe(self.on.upgrade_charm, self._on_upgrade_charm)
 
     def _on_config_changed(self, _: ops.ConfigChangedEvent) -> None:
-        """Handle changed configuration."""
-        self.unit.status = ops.ActiveStatus()
+        """Handle config changed."""
+        self.reconcile()
 
     def _on_install(self, _: ops.InstallEvent) -> None:
         """Handle install."""
-        self.unit.status = ops.MaintenanceStatus("Preparing irc bridge")
-        self.irc.prepare()
-        self.unit.status = ops.ActiveStatus()
+        self.reconcile()
 
     def _on_start(self, _: ops.StartEvent) -> None:
         """Handle start."""
-        self.irc.start()
-        self.unit.status = ops.ActiveStatus()
+        self.reconcile()
 
     def _on_stop(self, _: ops.StopEvent) -> None:
         """Handle stop."""
-        self.irc.stop()
+        self._irc.stop()
+
+    @property
+    def _charm_config(self) -> CharmConfig:
+        """Reconcile the charm.
+
+        Returns:
+            CharmConfig: The reconciled charm configuration.
+        """
+        return CharmConfig(
+            ident_enabled=self.model.config["ident_enabled"],
+            bot_nickname=self.model.config["bot_nickname"],
+            bridge_admins=self.model.config["bridge_admins"],
+        )
+
+    def reconcile(self) -> None:
+        """Reconcile the charm.
+
+        This is a more simple approach to reconciliation,
+        adapted from Charming Complexity sans state and observers.
 
-    def _on_upgrade_charm(self, _: ops.UpgradeCharmEvent) -> None:
-        """Handle upgrade-charm."""
-        self.unit.status = ops.MaintenanceStatus("Upgrading dependencies")
-        self.irc.prepare()
+        Being a simple charm, we don't need to do much here.
+
+        Ensure we have a database relation,
+        ensure we have a relation to matrix,
+        populate database connection string and matrix homeserver URL
+        in the config template and (re)start the service.
+        """
+        ops.MaintenanceStatus("Reconciling charm")
+        try:
+            logger.info("DB Reconciling charm")
+            db = self._database.uri
+        except ValidationError:
+            self.unit.status = ops.BlockedStatus("Database configuration not correct")
```

Comment:
I saw that reconcile is called during install event. If there is no database integration, would be catched by this try, right?

Reply:
It shouldn't cause a problem because of the holistic approach of that code, next event that matches all the requirements will install the snap. And we know an even will be triggered if the database is related, or the configuration changes